### PR TITLE
[ISSUE-2892] Sync and events refactor

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
issue: [2892](https://github.com/GetStream/stream-chat-android/issues/2892)

### 🎯 Goal

Refactor `EventHandlerImpl` to stop using `chatDomain` and handle events with `OfflinePlugin`

### 🛠 Implementation details

Removing many methods related to event handling from `ChatDomainImpl`.

### 🧪 Testing

Test events in the app and the sync with backend

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- ~[ ] Changelog is updated with client-facing changes~
- [x] New code is covered by unit tests. todo
- ~[ ] Comparison screenshots added for visual changes~
- ~[ ] Affected documentation updated (KDocs, docusaurus, tutorial)~

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

giphy
_So many failed attempts to create this PR D=_

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 1]
```diff
@@ -380,6 +380,7 @@ public class ChatClient internal constructor(
      */
     @CheckResult
     public fun connectUser(user: User, tokenProvider: TokenProvider): Call<ConnectionData> {
+        InitializationCoordinator.getOrCreate().userSet(user)
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:383]: Why do we need this? I think we are already doing it inside `setUser` 🤔

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt [hunk 2]
```diff
@@ -160,7 +146,7 @@ internal class ChatDomainImpl internal constructor(
     internal lateinit var repos: RepositoryFacade
 
     /** the event subscription */
-    private var eventSubscription: Disposable = EMPTY_DISPOSABLE
+    // private var eventSubscription: Disposable = EMPTY_DISPOSABLE
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt:149]: Can we remove this?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt [hunk 3]
```diff
@@ -57,68 +59,126 @@ import io.getstream.chat.android.client.models.ChannelCapabilities
...
 import io.getstream.chat.android.offline.repository.RepositoryFacade
 import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.launch
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import java.util.Date
+import java.util.InputMismatchException
 
 internal class EventHandlerImpl(
-    private val domainImpl: ChatDomainImpl,
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt:74]: 🎉

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt [hunk 4]
```diff
@@ -583,7 +607,28 @@ internal class EventHandlerImpl(
         }
     }
 
-    internal fun clear() {
-        firstConnect = true
+    private suspend fun updateCurrentUser(me: User) {
+        val currentUser = mutableGlobalState.user.value
+        if (me.id != currentUser?.id) {
+            throw InputMismatchException("received connect event for user with id ${me.id} while chat domain is configured for user with id ${currentUser?.id}. create a new ChatDomain when connecting a different user.")
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt:613]: We should update the message as we won't have chat domain :)

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerProvider.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package io.getstream.chat.android.offline.event
+
+internal object EventHandlerProvider {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerProvider.kt:3]: Can we add KDocs here?

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt [hunk 6]
```diff
@@ -152,6 +153,19 @@ internal class ChannelLogic(
             }
     }
 
+    internal fun state(): ChannelState {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt:156]: Let's add a KDoc here

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt [hunk 7]
```diff
@@ -152,6 +153,19 @@ internal class ChannelLogic(
...
+        return mutableState
+    }
+
+    internal suspend fun watch(messagesLimit: Int = 30, userPresence: Boolean) {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/channel/logic/ChannelLogic.kt:160]: And here

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/querychannels/logic/QueryChannelsLogic.kt [hunk 8]
```diff
@@ -90,6 +92,10 @@ internal class QueryChannelsLogic(
             .let { Result.success(it) }
     }
 
+    internal fun state(): QueryChannelsState {
```
Comment (adasiewiczr) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/experimental/querychannels/logic/QueryChannelsLogic.kt:95]: Let's add Kdoc

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt (hunk 2 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/ChatDomainImpl.kt:149]: Ops 🤦

**Pull request merged**